### PR TITLE
The great API clean up of 2016

### PR DESCRIPTION
**Pull request by decarbonization**
**Do not merge without two +1s**

This pull request is all about providing a more intuitive API for interacting with remote peripherals through the Buruberi library. List of changes includes:

- Dropped deprecated methods for interacting with characteristics through the `GattPeripheral` interface
- Dropped the old interface for checking if errors will cause instability
- Dropped device support whitelist – it does not belong in a low level BLE library
- Dropped the `Operation` enum intended for clients – it does not belong in a low level BLE library
- Added more annotations to help enforce proper usage of the library in clients
- Cleaned up project code
    + Introduced comments to clarify when a field/method is package scoped, or visible for testing
    + Introduced new, more specific interfaces for many things that used to use `Action*` and `Func*` interfaces from RxJava
    + Introduced some new companion classes to improve DRYness of project code
    + Improved structure of some classes
    + Removed dead code in project, and fixed lint issues
- Replaced the old peripheral-centric packet processing system with a new characteristic-centric one
- Added support for reading characteristic values
- Improved test coverage
- Updated to latest version of support libraries and build tools
- Rolled major project version over to `1.0.0`

### CODE REVIEW

**Comment by zafrani** [comment]
:+1:  Will go through it a second time today.

**Comment by decarbonization** [comment]
I can produce an APK that uses this refactored code for anyone interested in verifying it runs correctly

File: buruberi-core/src/main/java/is/hello/buruberi/bluetooth/stacks/android/GattDispatcher.java [hunk 1]
```diff
@@ -29,85 +29,44 @@
...
     private final Handler dispatcher = new Handler(Looper.getMainLooper());
 
-    @Nullable GattPeripheral.PacketHandler packetHandler;
-    @Nullable Action2<BluetoothGatt, Integer> onServicesDiscovered;
-    @Nullable Action3<BluetoothGatt, BluetoothGattCharacteristic, Integer> onCharacteristicWrite;
-    @Nullable Action3<BluetoothGatt, BluetoothGattDescriptor, Integer> onDescriptorWrite;
+    private final List<ConnectionListener> connectionStateListeners = new ArrayList<>();
+    /*package*/ @Nullable CharacteristicReadListener characteristicRead;
+    /*package*/ @Nullable ServicesDiscoveredListener servicesDiscovered;
+    /*package*/ @Nullable CharacteristicWriteListener oharacteristicWrite;
```
Comment (jimmymlu) [inline buruberi-core/src/main/java/is/hello/buruberi/bluetooth/stacks/android/GattDispatcher.java]: `o` != `c`, but maybe in some other language?

File: buruberi-core/src/main/java/is/hello/buruberi/bluetooth/stacks/android/NativeGattCharacteristic.java [hunk 2]
```diff
@@ -89,30 +95,80 @@ public int getDescriptorPermissions(@NonNull UUID identifier) {
...
+
+                        gattDispatcher.characteristicRead = null;
+                        peripheral.removeDisconnectListener(onDisconnect);
+                    }
+                };
+                if (gatt.readCharacteristic(wrappedCharacteristic)) {
+                    timeout.schedule();
+                } else {
+                    gattDispatcher.characteristicRead = null;
+                    subscriber.onError(new GattException(BluetoothGatt.GATT_FAILURE,
```
Comment (jimmymlu) [inline buruberi-core/src/main/java/is/hello/buruberi/bluetooth/stacks/android/NativeGattCharacteristic.java]: as mentioned, maybe removeDisconnectListener here?

File: buruberi-core/src/main/java/is/hello/buruberi/bluetooth/stacks/android/NativeGattPeripheral.java [hunk 3]
```diff
@@ -678,25 +712,26 @@ public void call(BluetoothGatt gatt, Integer status) {
                 if (gatt.discoverServices()) {
                     timeout.schedule();
                 } else {
-                    gattDispatcher.onServicesDiscovered = null;
+                    gattDispatcher.servicesDiscovered = null;
```
Comment (jimmymlu) [inline buruberi-core/src/main/java/is/hello/buruberi/bluetooth/stacks/android/NativeGattPeripheral.java]: remove disconnect listener?

**Comment by jimmymlu** [comment]
:+1:

**Comment by jimmymlu** [comment]
@zafrani are you still a :+1: on this?  if so, we can merge this once CI is done

**Pull request merged**